### PR TITLE
Laboratory work #1, Kozlova Anastasia - 19FPL2

### PR DESCRIPTION
**Pull request by AnastasiaKozlova**

### CODE REVIEW

File: lab_1/main.py [hunk 1]
```diff
@@ -3,16 +3,23 @@
...
+                    word += i
+                else:
+                    if word:
+                        print(word.lower())
+
+                        word = ''
+tokenize()
+pass
 
-def tokenize(text: str) -> list:
```
Comment (WhiteJaeger) [inline lab_1/main.py]: Please, leave the description of the function how it is - use the already written template.

File: lab_1/main.py [hunk 2]
```diff
@@ -3,16 +3,23 @@
 A concordance extraction
 """
 
+import re
+def tokenize():
+    word = ''
+    f = open('data.txt', 'r')
```
Comment (WhiteJaeger) [inline lab_1/main.py]: You do not need to open it here: there is a function `read_from_file` which can do just that.

File: lab_1/main.py [hunk 3]
```diff
@@ -1,31 +1,27 @@
-"""
-Lab 1
-A concordance extraction
-"""
+num = "1 2 3 4 5 6 7 8 9 0"
+punc = " ',:@.-()/"
+import re
+def tokenize(text: str) -> list:
+    if type(text) != str: #проверяем, строка ли это
```
Comment (WhiteJaeger) [inline lab_1/main.py]: `type` is considered a bad style.
Try using `isinstance` for type checking.

File: lab_1/main.py [hunk 4]
```diff
@@ -1,31 +1,27 @@
...
+        return []
+ for i in punc:
```
Comment (WhiteJaeger) [inline lab_1/main.py]: Indentation is wrong here, it is inside a function.

File: lab_1/main.py [hunk 5]
```diff
@@ -1,31 +1,27 @@
...
+    text = text.replace(i, ' ') #убираем знаки препинания
+    text = text.lower() #переводим текст в нижний регистр
```
Comment (WhiteJaeger) [inline lab_1/main.py]: Do you really need to do it every punctuation mark?

File: lab_1/main.py [hunk 6]
```diff
@@ -1,31 +1,27 @@
...
-    :param tokens: a list of tokens
-    :param stop_words: a list of stop words
-    :return: a list of tokens without stop words
-    e.g. tokens = ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
-    stop_words = ['the', 'is']
-    --> ['weather', 'sunny', 'man', 'happy']
-    """
-    pass
+    if isinstance(tokenize_list, list) and isinstance(stop_words, list): #проверяем, списки ли это
+        for i in stop_words:
```
Comment (WhiteJaeger) [inline lab_1/main.py]: `i` is not very descriptive, right?

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_1/main.py [hunk 7]
```diff
@@ -1,42 +1,35 @@
...
-A concordance extraction
-"""
+import re
+def tokenize(text: str) -> list:
+    if isinstance(text,str): #проверяем, строка ли это
+        return []
+    text = re.sub(r'[^A-Za-z 0-9]', '', text) # с помощью рег. выраж. убираю знаки препинания и числительные
+    tokenize_list = text.split().lower()
+    return tokenize_list
+pass
```
Comment (WhiteJaeger) [inline lab_1/main.py]: Remove as it was a placeholder.

File: lab_1/main.py [hunk 8]
```diff
@@ -1,42 +1,35 @@
...
-    :param text: the initial text
-    :return: a list of lowercased tokens without punctuation
-    e.g. text = 'The weather is sunny, the man is happy.'
-    --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
-    """
-    pass
 
 
 def remove_stop_words(tokens: list, stop_words: list) -> list:
-    """
```
Comment (WhiteJaeger) [inline lab_1/main.py]: Please, leave the description of the function how it is - use the already written template.

File: lab_1/main.py [hunk 9]
```diff
@@ -1,42 +1,35 @@
...
-    Removes stop words
-    :param tokens: a list of tokens
-    :param stop_words: a list of stop words
-    :return: a list of tokens without stop words
-    e.g. tokens = ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
-    stop_words = ['the', 'is']
-    --> ['weather', 'sunny', 'man', 'happy']
-    """
-    pass
+    if isinstance(tokenize_list, list) and isinstance(stop_words, list): #проверяем, списки ли это
```
Comment (WhiteJaeger) [inline lab_1/main.py]: Use your structure from the first function - get rid of nesting.

File: lab_1/main.py [hunk 10]
```diff
@@ -1,42 +1,35 @@
...
 
 def calculate_frequencies(tokens: list) -> dict:
-    """
-    Calculates frequencies of given tokens
-    :param tokens: a list of tokens without stop words
-    :return: a dictionary with frequencies
-    e.g. tokens = ['weather', 'sunny', 'man', 'happy']
-    --> {'weather': 1, 'sunny': 1, 'man': 1, 'happy': 1}
-    """
-    pass
```
Comment (WhiteJaeger) [inline lab_1/main.py]: Please, leave the description of the function how it is - use the already written template.

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_1/main.py [hunk 11]
```diff
@@ -1,55 +1,84 @@
...
 def tokenize(text: str) -> list:
     """
-    Splits sentences into tokens, converts the tokens into lowercase, removes punctuation
-    :param text: the initial text
-    :return: a list of lowercased tokens without punctuation
-    e.g. text = 'The weather is sunny, the man is happy.'
-    --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
-    """
-    pass
+        Splits sentences into tokens, converts the tokens into lowercase, removes punctuation
```
Comment (WhiteJaeger) [inline lab_1/main.py]: Different indentation from the original one.

File: lab_1/main.py [hunk 12]
```diff
@@ -1,55 +1,84 @@
-"""
-Lab 1
-A concordance extraction
```
Comment (WhiteJaeger) [inline lab_1/main.py]: Please, leave the description as it is.

File: lab_1/main.py [hunk 13]
```diff
@@ -1,55 +1,84 @@
...
+        Splits sentences into tokens, converts the tokens into lowercase, removes punctuation
+        :param text: the initial text
+        :return: a list of lowercased tokens without punctuation
+        e.g. text = 'The weather is sunny, the man is happy.'
+        --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
+        """
+    if not isinstance(text, str):
+        return []
+
+    else:
```
Comment (WhiteJaeger) [inline lab_1/main.py]: No need for `else`-block.

File: lab_1/main.py [hunk 14]
```diff
@@ -1,55 +1,84 @@
...
-    e.g. tokens = ['weather', 'sunny', 'man', 'happy']
-    --> {'weather': 1, 'sunny': 1, 'man': 1, 'happy': 1}
-    """
-    pass
+        Calculates frequencies of given tokens
+        :param tokens: a list of tokens without stop words
+        :return: a dictionary with frequencies
+        e.g. tokens = ['weather', 'sunny', 'man', 'happy']
+        --> {'weather': 1, 'sunny': 1, 'man': 1, 'happy': 1}
+        """
```
Comment (WhiteJaeger) [inline lab_1/main.py]: Different indentation from the original one.

File: lab_1/main.py [hunk 15]
```diff
@@ -1,55 +1,84 @@
...
+    if not isinstance(tokens, list):
+        return {}
+    dict = {} #создаем словарь
+    if isinstance(tokens[0], str):
```
Comment (WhiteJaeger) [inline lab_1/main.py]: why only the first element?

File: lab_1/main.py [hunk 16]
```diff
@@ -1,55 +1,84 @@
...
-    """
-    pass
+        Returns the most common words
+        :param freq_dict: a dictionary with frequencies
+        :param top_n: a number of the most common words to return
+        :return: a list of the most common words
+        e.g. tokens = ['weather', 'sunny', 'man', 'happy', 'and', 'dog', 'happy']
+        top_n = 1
+        --> ['happy']
+        """
```
Comment (WhiteJaeger) [inline lab_1/main.py]: Different indentation from the original one.

File: lab_1/main.py [hunk 17]
```diff
@@ -87,16 +135,29 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
     right_n = 3
     --> [['man', 'is'], ['dog, 'cat']]
     """
-    pass
+    if not isinstance(tokens, list) and not isinstance(word, str) and not isinstance(left_n, int) and not isinstance(right_n, int)\
+    and not (left_n < 1) and not (right_n < 1):
+        return []
+    previous_conc = get_concordance(tokens, word, left_n, right_n)
+    ad_word = []
+    for cotext in previous_conc:
```
Comment (WhiteJaeger) [inline lab_1/main.py]: typo: `context`

File: lab_1/main.py [hunk 18]
```diff
@@ -105,7 +166,10 @@ def write_to_file(path_to_file: str, content: list):
     """
     Writes the result in a file
     """
-    pass
+    with open(path_to_file, 'r', encoding='utf-8') as concordance_file:
+        for i in content:
```
Comment (WhiteJaeger) [inline lab_1/main.py]: `i` is not very descriptive.

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_1/main.py (hunk 15 above)
Comment (AnastasiaKozlova) [reply lab_1/main.py]: if the first element is a string it doesn't give us a dictionary with the type of None

Comment (WhiteJaeger) [reply lab_1/main.py]: But what if `tokens[0]` is a `str`, but `tokens[1]` is `None`?

File: lab_1/main.py [hunk 19]
```diff
@@ -105,7 +165,10 @@ def write_to_file(path_to_file: str, content: list):
     """
     Writes the result in a file
     """
-    pass
+    with open(path_to_file, 'r', encoding='utf-8') as concordance_file:
+        for cont in content:
+            return concordance_file.write('/n'. join(cont))
```
Comment (WhiteJaeger) [inline lab_1/main.py]: You should not use `return` here. Just `write`.

**Review by WhiteJaeger — APPROVED** [review]
Great work! You are done here.
Just fix the remaining comments.

**Comment by WhiteJaeger** [comment]
Open another PR with lab_2.

**Pull request closed without merging**